### ISSUE TITLE
no-undef not reported in switch case block

### ISSUE THREAD

**Issue by lwr** [eslint/eslint#10820]
**Tell us about your environment**
* **ESLint Version:** 4.19.1
* **Node Version:** 10.7.0
* **npm Version:** 6.4.0

**What parser (default, Babel-ESLint, etc.) are you using?** default

**Please show your full configuration:**
- https://github.com/Mailtech/eslint-config-coremail/blob/0.3.0/config/standard.yaml

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
switch (1) {
case 0:
    const x = 1;
    break;
case 1:
    console.log(x);
    break;
}
```

```bash
eslint test.js
```

**What did you expect to happen?**
> error  'x' is not defined  no-undef

**What actually happened? Please include the actual, raw output from ESLint.**
no error

**Comment by platinumazure** [comment]
Case statements don't introduce a new scope, so from the perspective of `no-undef` this is working as designed. That said, I would be curious to see if we can improve this in some way...

Hoping someone else from the team might have some thoughts.

**Comment by lwr** [comment]
@platinumazure there might be an other more difficult case
```js
let x = 1, y = 1;
    switch (1) {
    case 0:
        let y = 2;
        break;
    case 1:
        console.log(x); // OK
        console.log(y); // ReferenceError: y is not defined
        break;
    }
}
```
case statement do not introduce a new scope, but switch statement do, so although y is defined but it is hidden in switch block and should be undefined.

**Comment by kaicataldo** [comment]
Can someone explain why the second example behaves as it does? Just tested it out, and if you remove the `y` declaration in the first case (`0`), it executes as it should. But as as soon as you add that second declaration the error described in the comment above occurs.

**Comment by not-an-aardvark** [comment]
This is intended to be caught by [`no-case-declarations`](https://eslint.org/docs/rules/no-case-declarations). The ReferenceErrors are resulting from a temporal dead zone reference to `y`, because `y` is declared in scope but is read before the declaration is executed.

**Comment by kaicataldo** [comment]
Interesting! I didn't realize the declaration would still be evaluated in scope if a given case it was declared in wasn't executed. TIL, thanks!

**Comment by lwr** [comment]
Seems duplicated with #4278

But I agreed it would be much better if we can have some enhancement of no-undef or no-use-before-define to catches TDZ error

It's ugly to wrap case / default statements into blocks

**Comment by nzakas** [comment]
The team has determined that this behavior is working as intended, so closing this issue.
